client/txn.go

```diff
@@ -0,0 +1,135 @@
+package client
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+	xlineapi "github.com/xline-kv/go-xline/api/xline"
+)
+
+type Txn interface {
+	// If takes a list of comparison. If all comparisons passed in succeed,
+	// the operations passed into Then() will be executed. Or the operations
+	// passed into Else() will be executed.
+	If(cs ...Cmp) Txn
+
+	// Then takes a list of operations. The Ops list will be executed, if the
+	// comparisons passed in If() succeed.
+	Then(ops ...Op) Txn
+
+	// Else takes a list of operations. The Ops list will be executed, if the
+	// comparisons passed in If() fail.
+	Else(ops ...Op) Txn
+
+	// Commit tries to commit the transaction.
+	Commit() (*TxnResponse, error)
+}
+
+type txn struct {
+	// Name of the kv client, which will be used in CURP propose id generation
+	name string
+	// The client running the CURP protocol, communicate with all servers.
+	curpClient curpClient
+	// The auth token
+	token string
+
+	mu    sync.Mutex
```

Comment:
What is this lock protecting? And what contention should this lock preventing from?

Reply:
Should we provide thread-safety for client to build a txn concurrently? How does etcd-client handle this?

Reply:
I think we need to. etcd did the same thing.